### PR TITLE
AVX2 blending

### PR DESCRIPTION
**Pull request by janisozaur**
AVX2 hype.

This is preliminary support for some AVX2 code. It is best pronounced in debug builds (duh) and on skylake+ (_looking at you @Gymnasiast and @IntelOrca_).

`benchgfx ../test/tests/testdata/bpb.sv6` results:
Skylake, `Debug`, went down from 19.42s to 18.05s.
Skylake, `RelWithDebInfo`, went down from 8.12s to 7.65s
Haswell, `Debug`, went down from 13.72s to 12.64s.

I am aware this will break non-AVX2 CPUs, I still need to do run-time AVX2 selection with some fallback. It is also not yet exactly correct, some pro- and epilogue may be required.

### CODE REVIEW

File: src/openrct2/drawing/AVX2Drawing.cpp [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+    else
+    {
+        mask_scalar(width, height, maskSrc, colourSrc, dst, maskWrap, colourWrap, dstWrap);
+    }
+}
+
+#else
+
+#ifdef OPENRCT2_X86
+#error You have to compile this file with AVX2 enabled, when targetting x86!
```
Comment (Gymnasiast) [inline src/openrct2/drawing/AVX2Drawing.cpp]: targeting

File: src/openrct2/drawing/drawing.c [hunk 2]
```diff
@@ -476,7 +476,12 @@ void (*mask_fn)(sint32 width, sint32 height, const uint8 * RESTRICT maskSrc, con
 
 void mask_init()
 {
-    if (sse41_available())
+    if (avx2_available())
+    {
+        log_verbose("registering AVX2 mask function");
+        mask_fn = mask_avx2;
+    }
+	else if (sse41_available())
```
Comment (Gymnasiast) [inline src/openrct2/drawing/drawing.c]: A tab!!!!!1!

**Comment by IntelOrca** [comment]
This needs rebasing.

To fix MSVC, just append `__AVX2__` to the list of defined preprocessors in `common.props`.

File: src/openrct2/util/Util.cpp [hunk 3]
```diff
@@ -188,10 +188,10 @@ sint32 bitscanforward(sint32 source)
 #endif
 
 #ifdef OPENRCT2_X86
-static bool cpuid_x86(uint32 * cpuid_outdata, sint32 eax)
+static bool cpuid_x86(uint32 * cpuid_outdata, sint32 eax, sint32 ecx = 0)
 {
 #if defined(OpenRCT2_CPUID_GNUC_X86)
-    int ret = __get_cpuid(eax, &cpuid_outdata[0], &cpuid_outdata[1], &cpuid_outdata[2], &cpuid_outdata[3]);
+    int ret = __get_cpuid_count(eax, ecx, &cpuid_outdata[0], &cpuid_outdata[1], &cpuid_outdata[2], &cpuid_outdata[3]);
```
Comment (janisozaur) [inline src/openrct2/util/Util.cpp]: Ugh, they added a new call that allows specifying subleaf in recent GCC version, but removed the legacy behaviour which specified subleafs for the user.

I think I'll have to change the avx2 detection to use GCC's builtin.

File: src/openrct2/util/Util.cpp [hunk 4]
```diff
@@ -218,12 +218,20 @@ bool sse41_available()
 bool avx2_available()
 {
 #ifdef OPENRCT2_X86
+// For GCC and similar use the builtin function, as cpuid changed its semantics in
+// https://github.com/gcc-mirror/gcc/commit/132fa33ce998df69a9f793d63785785f4b93e6f1
+// which causes it to ignore subleafs, but the new function is unavailable on Ubuntu's
+// prehistoric toolchains
+#if defined(OpenRCT2_CPUID_GNUC_X86)
+    return __builtin_cpu_supports("avx2");
```
Comment (janisozaur) [inline src/openrct2/util/Util.cpp]: Great, now I'm running into https://bugs.launchpad.net/ubuntu/+source/gcc-5/+bug/1568899…

File: src/openrct2/CMakeLists.txt [hunk 5]
```diff
@@ -106,6 +106,13 @@ if (NOT DISABLE_NETWORK)
     endif ()
 endif ()
 
+if (NOT APPLE AND NOT MINGW)
+    # This is ugly hack to work around https://bugs.launchpad.net/ubuntu/+source/gcc-5/+bug/1568899.
+    # Once C++17 is enabled (and thus old compilers are no longer supported, this needs to be gone.
+    # We cannot simply detect the _compiler_ version, as the bug exists with the C++ _library_
+    target_link_libraries(${PROJECT} gcc_s gcc)
```
Comment (IntelOrca) [inline src/openrct2/CMakeLists.txt]: GCC's turn to have a bug now :)

Comment (janisozaur) [reply src/openrct2/CMakeLists.txt]: Yes, GCC 5, first released 2015-04-22 and already fixed in newer versions.

**Comment by marijnvdwerf** [comment]
| | `develop` | PR
|--|--|--|
| Debug| 16,17  | 15,95|
| Release | 6,64 | 6,72|

**Pull request merged**